pkg/sanity/node.go

```diff
@@ -449,6 +449,67 @@ var _ = DescribeSanity("Node Service", func(sc *TestContext) {
 			Expect(ok).To(BeTrue())
 			Expect(serverError.Code()).To(Equal(codes.InvalidArgument), "unexpected error: %s", serverError.Message())
 		})
+
+		It("should fail if target path was not removed by driver", func() {
+			if !nodeVolumeStatsSupported {
+				Skip("NodeGetVolumeStats not supported")
+			}
+
+			name := UniqueString("sanity-node-unpublish-volume")
+			vol := createVolume(name)
+			volid := vol.GetVolume().GetVolumeId()
+			volpath := sc.TargetPath + "/target"
+
+			By("Getting a node id")
+			nid, err := r.NodeGetInfo(
+				context.Background(),
+				&csi.NodeGetInfoRequest{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(nid).NotTo(BeNil())
+			Expect(nid.GetNodeId()).NotTo(BeEmpty())
+
+			By("Staging and publishing a volume")
+			conpubvol := controllerPublishVolume(name, vol, nid)
+			_ = nodeStageVolume(name, vol, conpubvol)
+			_ = nodePublishVolume(name, vol, conpubvol)
+
+			// Verify that NodeGetVolumeStats is successful
+			// before the NodeUnpublishVolume call
+			By("Getting node volume stats")
+			_, err = r.NodeGetVolumeStats(
+				context.Background(),
+				&csi.NodeGetVolumeStatsRequest{
+					VolumeId:   volid,
+					VolumePath: volpath,
+				},
+			)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Unpublishing the volume")
+			_, err = r.NodeUnpublishVolume(
+				context.Background(),
+				&csi.NodeUnpublishVolumeRequest{
+					VolumeId:   volid,
+					TargetPath: volpath,
+				},
+			)
+			Expect(err).NotTo(HaveOccurred())
+
+			// NodeGetVolumeStats is expected to return NotFound
+			// if the TargetPath is removed by NodeUnpublishVolume
+			By("Checking NodeGetVolumeStats returns NotFound error")
```

Comment:
This seems to make assumptions about how `NodeGetVolumeStats` is implemented. It could return an error because it tracks volume state and knows that `NodeUnpublishVolume` was called even when that function didn't remove the target path.
 
What you really need is to check on the node whether the directory is gone. Perhaps you can use `CreateTargetDir` (https://github.com/kubernetes-csi/csi-test/blob/a251c44fd49d9eedd55a40b71e5da4ad080ba431/pkg/sanity/sanity.go#L124) or whatever is configured (https://github.com/kubernetes-csi/csi-test/blob/a251c44fd49d9eedd55a40b71e5da4ad080ba431/pkg/sanity/sanity.go#L143). At least the internal implementation (https://github.com/kubernetes-csi/csi-test/blob/a251c44fd49d9eedd55a40b71e5da4ad080ba431/pkg/sanity/sanity.go#L362-L367) fails when the directory exists.

But that looks like a hack. A cleaner solution is to introduce a `CheckPath` feature (with a configurable callback and command) which returns "file", "directory", "other" and "not_found", or an error, and then use that here.

Reply:
Thanks for the feedback @pohly, this was very helpful.
I implemented the CheckPath function, but with a couple of minor differences from what you described.
1) I used return codes instead of strings, but it does return the 4 conditions you described. The test in node.go only checks for found or not found though. Since the spec states "file or directory" I did not want to make assumptions about the file type in the sanity test.
2) I only did the configurable callback, not the command string that can override that behavior like in createMountTargetLocation() and removeMountTargetLocation(). This simplified the implementation, and I question whether the command string would be as useful in this context as it is for create / remove commands. Translating those 4 conditions into error codes (or strings) would add some complexity to the command. If the command string supports an essential use case though, I can add in.

Reply:
The command string is essential for those users who invoke the `csi-sanity` command to run tests. They can't extend that binary with Go callbacks.

Reply:
Thanks for elaborating on this, I added the command string option.
I tested it with the following `check_path.sh` bash script:
```bash
#!/bin/bash
if [ -f "$1" ]; then
        echo "1" # file
elif [ -d "$1" ]; then
        echo "2" # dir
elif [ ! -e "$1" ]; then
        echo "3" # not found
else
        echo "4" # other
fi
```
And that passes against the mock driver using the following args:
```
$ csi-sanity -csi.checkpathcmd ./check_path.sh -csi.endpoint /tmp/csi.sock
```

Reply:
Why use numbers as output? I find strings as in my original proposal easier to use because there's no chance that someone, say, returns 3 when they meant 4. The same strings can be string constants in Go for the Go callback, so it could be even made type safe there.

Reply:
You're right, strings are a better approach here. When I added the `Atoi` call I started to question if that was really the best option. Updated.